Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrUniqIdProvider.kt

```diff
@@ -51,11 +47,49 @@ internal class StubIrUniqIdProvider {
                 ?: error("Unexpected origin ${typeAlias.origin} for typealias ${typeAlias.alias.fqName}.")
     }.toUniqId()
 
+    private fun InteropMangler.uniqSymbolNameForClass(origin: StubOrigin): String? = when (origin) {
+        is StubOrigin.ObjCClass -> if (origin.isMeta) {
+            origin.clazz.metaClassUniqueSymbolName
+        } else {
+            origin.clazz.uniqueSymbolName
+        }
+        is StubOrigin.ObjCProtocol -> if (origin.isMeta) {
+            origin.protocol.metaClassUniqueSymbolName
+        } else {
+            origin.protocol.uniqueSymbolName
+        }
+        is StubOrigin.Struct -> origin.struct.uniqueSymbolName
+        is StubOrigin.Enum -> origin.enum.uniqueSymbolName
+        else -> null
+    }
+
+    fun uniqIdForClass(classStub: ClassStub): UniqId = with (mangler) {
+        when (classStub) {
+            is ClassStub.Simple,
+            is ClassStub.Enum -> uniqSymbolNameForClass(classStub.origin)
+            is ClassStub.Companion -> "${context.prefix}#Companion"
+        } ?: error("Unexpected origin ${classStub.origin} for enum class ${classStub.classifier.fqName}.")
+    }.toUniqId()
+
+    fun uniqIdForConstructor(constructorStub: ConstructorStub): UniqId =
+            "${context.prefix}#constructor#${constructorStub.mangleParameters()}".toUniqId()
+
     /**
      * MSB should be set to 1 for public declarations.
      * @see org.jetbrains.kotlin.ir.util.UniqId
      */
     private fun Long.markAsPublic() = this or (1L shl 63)
 
     private fun String.toUniqId() = UniqId(cityHash64().markAsPublic())
+
+    private fun FunctionalStub.mangleParameters(): String = parameters.joinToString(
```

Comment:
This method is kinda harmful. It is too general, and doesn't handle actual language-specific overloads (e.g. for Objective-C constructors).
Is it possible to avoid having such a fallback?

Reply:
Agree, it's the most controversial part of the PR. I'll try to rethink the implementation.